core/dbt/artifacts/resources/v1/semantic_model.py

```diff
@@ -42,7 +42,7 @@ class NodeRelation(dbtClassMixin):
     alias: str
     schema_name: str  # TODO: Could this be called simply "schema" so we could reuse StateRelation?
     database: Optional[str] = None
-    relation_name: Optional[str] = None
+    relation_name: Optional[str] = ""
```

Comment:
This PR is focused on tests and fixtures for tests. Was the default value of `relation_name` changed to make testing easier, or was there a functional purpose for the change?

Reply:
So after this got fixed we started seeing [test failures](https://github.com/dbt-labs/dbt-core/actions/runs/8713131767/job/23900944831?pr=9930) that take the form of
```
______________________ TestSemanticManifest.test_validate ______________________

self = <tests.unit.contracts.graph.test_semantic_manifest.TestSemanticManifest object at 0x7f52df26c040>
manifest = Manifest(nodes={'model.test.metricflow_time_spine': ModelNode(database='dbt', schema='analytics', name='metricflow_tim...d_path_count=0, static_analysis_path_count=0), _lock=<Lock(owner=None)>, _macros_by_name=None, _macros_by_package=None)

    def test_validate(self, manifest):
        sm_manifest = SemanticManifest(manifest)
>       assert sm_manifest.validate()

tests/unit/contracts/graph/test_semantic_manifest.py:28: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
core/dbt/contracts/graph/semantic_manifest.py:42: in validate
    semantic_manifest = self._get_pydantic_semantic_manifest()
core/dbt/contracts/graph/semantic_manifest.py:69: in _get_pydantic_semantic_manifest
    PydanticSemanticModel.parse_obj(semantic_model.to_dict())
.tox/unit/lib/python3.8/site-packages/pydantic/v1/main.py:526: in parse_obj
    return cls(**obj)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

__pydantic_self__ = PydanticSemanticModel()
data = {'config': {'enabled': True, 'group': None, 'meta': {}}, 'created_at': 1713305259.7969077, 'defaults': None, 'depends_on': {'macros': [], 'nodes': []}, ...}
values = {'defaults': None, 'description': 'Customer entity', 'dimensions': [], 'entities': [], ...}
fields_set = {'defaults', 'description', 'dimensions', 'entities', 'label', 'measures', ...}
validation_error = ValidationError(model='PydanticSemanticModel', errors=[{'loc': ('node_relation', 'relation_name'), 'msg': 'none is not an allowed value', 'type': 'type_error.none.not_allowed'}])

    def __init__(__pydantic_self__, **data: Any) -> None:
        """
        Create a new model by parsing and validating input data from keyword arguments.
    
        Raises ValidationError if the input data cannot be parsed to form a valid model.
        """
        # Uses something other than `self` the first arg to allow "self" as a settable attribute
        values, fields_set, validation_error = validate_model(__pydantic_self__.__class__, data)
        if validation_error:
>           raise validation_error
E           pydantic.v1.error_wrappers.ValidationError: 1 validation error for PydanticSemanticModel
E           node_relation -> relation_name
E             none is not an allowed value (type=type_error.none.not_allowed)

.tox/unit/lib/python3.8/site-packages/pydantic/v1/main.py:341: ValidationError
=============================== warnings summary ===============================
tests/unit/test_query_headers.py:44
  /home/runner/work/dbt-core/dbt-core/tests/unit/test_query_headers.py:44: DeprecationWarning: invalid escape sequence \/
    self.assertTrue(re.match(f"^\/\*.*\*\/\n{self.query}$", sql))  # noqa: [W605]

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform linux, python 3.8.18-final-0 -----------
Coverage XML written to file coverage.xml

=========================== short test summary info ============================
FAILED tests/unit/contracts/graph/test_semantic_manifest.py::TestSemanticManifest::test_validate
```

Reply:
The test that failed, `TestSemanticManifest.test_validate`, is a new test added by this PR. The question becomes, is this an error in the code or the test.

Reply:
It looks like the protocol for [NodeRelation](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/e4f029b29c2baab1acee4b9d0dc915a5574525c4/dbt_semantic_interfaces/protocols/semantic_model.py#L39) does have `node_relation` as a required string. The [PydanticNodeRelation](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/e4f029b29c2baab1acee4b9d0dc915a5574525c4/dbt_semantic_interfaces/implementations/semantic_model.py#L31) has a specific pydantic validator for handling [node_relation](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/e4f029b29c2baab1acee4b9d0dc915a5574525c4/dbt_semantic_interfaces/implementations/semantic_model.py#L41) 🤔 So why do we in core have it as optional

Reply:
Haha yes, maybe this does deserve a separate PR?

Reply:
Updating the default value of `relation_name` to `""` in the [class definition](https://github.com/dbt-labs/dbt-core/blob/4c1d0e92cd0a9c09659cc432c016dbc966b89daf/core/dbt/artifacts/resources/v1/semantic_model.py#L41) should be fine. The `NodeRelation` class is only instantiated on a `SemanticModel` [in one place](https://github.com/dbt-labs/dbt-core/blob/4c1d0e92cd0a9c09659cc432c016dbc966b89daf/core/dbt/parser/manifest.py#L1178), and it passes in a value for `relation_name`.